### ISSUE TITLE
[JSP] NoClassAttribute rule is generating a false positive.

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2033]
**Affects PMD Version:** 
6.18.0

**Rule:**
NoClassAttribute.

**Description:**
The rule generates an error for a jsp:usebean tag where one of the attributes is named "class".

**Code Sample demonstrating the issue:**
In JSP page:
```jsp
<jsp:usebean id="data" class="jdbc.JdbcBean"/>
```

**Running PMD through:
Any.

**Comment by Fernal73** [comment]
What is the motivation for this rule? Is it to ensure that JSF tags use the styleclass attribute as against the class attribute for an HTML element? Or is to discourage in-lining of style elements in an HTML element, which is a best practice recommended for HTML? I'd have to assume it's the former. Either way, the above is still a bug.

**Comment by adangel** [comment]
> What is the motivation for this rule? Is it to ensure that JSF tags use the styleclass attribute as against the class attribute for an HTML element? Or is to discourage in-lining of style elements in an HTML element, which is a best practice recommended for HTML? I'd have to assume it's the former. Either way, the above is still a bug.

Quoting the rule description:

> Do not use an attribute called ‘class’. Use "styleclass" for CSS styles.

So, this rule indeed seems to ensure, that JSF tags only use the "styleClass" attribute instead of the directly the html attribute "class". (Note, there is no "styleclass" attribute with all lowercase letters...)
There are unfortunately no good test cases (see https://github.com/pmd/pmd/blob/master/pmd-jsp/src/test/resources/net/sourceforge/pmd/lang/jsp/rule/bestpractices/xml/NoClassAttribute.xml), e.g. no test case, that makes sure, that the alleged valid case with styleclass doesn't raise a violation (so protection against a false-positive even for the case the rule was written for).

The rule could be easily adjusted to exclude elements from the "jsf" namespace, e.g.

```
//Element[@HasNamespacePrefix = false() or @NamespacePrefix != 'jsp']
    /Attribute[ upper-case(@Name)="CLASS" ]
```
(note: xpath 2.0 syntax)

The valid namespaces should then be supplied via a property, so that it can be adjusted by the rule users.

However, I can't say, whether this rule makes sense at all. E.g. the standard tag libraries ([h.outputText](https://docs.oracle.com/javaee/6/javaserverfaces/2.1/docs/vdldocs/facelets/h/outputText.html) don't support a class attribute anyway, you have to use styleClass.

Also quoting from your linked stackoverflow question:

> And class attribute is used with HTML elements, and can't be used directly with JSF components.

Maybe the rule wanted to prevent using "style"? Unless it wanted to prevent that html is directly used and every rendering should be done through jsf components (or similar alternatives), but than the rule should have been named "NoHtml"...

**Comment by Fernal73** [comment]
https://blog.tbhcreative.com/2015/08/10-best-practices-in-html.html?m=1

**Comment by linusjf** [comment]
Another point that's unrelated, is PMD, at some point , going to introduce a feature that allows users to track rules and the metrics associated with each rule as to how many times it occurs in the code base? This would be internal with an option to enable sharing that data with PMD via some sort of web service. Just a thought.

This would be kind of like the service Gradle provides to share builds via its scan flag. It's optional but it would help garner statistics of PMD runs.

**Comment by adangel** [comment]
> You kinda got me with that Xpath expression that amends the rule. Is there a central location 
> where these Xpath expressions for each individual rule is located so that they can be studied 
> individually without firing up the Rule Designer?

The xpath expressions for the built in rules are documented in the rule documentation. In the sources, they are in the corresponding category rulesets, e.g.
https://github.com/pmd/pmd/blob/9923dd28de4da1c5c5f92c748e9d63aa5c5be3df/pmd-jsp/src/main/resources/category/jsp/bestpractices.xml#L60



> Another point that's unrelated, is PMD, at some point , going to introduce a feature that allows 
> users to track rules and the metrics associated with each rule as to how many times it occurs in 
> the code base? This would be internal with an option to enable sharing that data with PMD via 
> some sort of web service. Just a thought.

Maybe. I guess that requires some sort of "multi file analysis" then, since you want to get an overview of the whole project. Currently the rules operate only on single files and don't see the big picture. For certain metrics, this is enough, but for others (e.g. calculating the average class size and compare each class to this average) we need this multi file analysis. Reporting these metrics then another point: Currently PMD is a standalone application, so we would probably extend/add report formats for metrics.

**Comment by linusjf** [comment]
https://docs.pylint.org/en/1.6.0/output.html#reports-section

Pylint's reports are quite impressive.

What I'd really like is a way to send an error report to a PMD web service
that logs the error information from a PMD error with the user only having
to say yes or no. It should default to No. But there could be a way to
configure this to Yes by the user. Would that be too intrusive?

Why wouldn't users want an overview of the whole project? I, for one, would be very interested in knowing which rule is violated most. Is it a pointer to inadequate training or coding skills or that the rule is irrelevant or unworkable (as it is) and can be dropped or needs to be modified? 
What are the other advantages of having an overview?
A grouping of number of  rules violated by severity levels would also be useful. 
Which metrics would be more useful?

This ought to be tackled in a separate issue. 

Can you go ahead and write up the issues if you think they're worth implementing? What's the criteria for triaging them, anyway? The impression I have is that issues are taken up by volunteers interested and willing to implement them after they're ? Is there any other methodology followed? If you subscribed to Github sponsors, then it's possible that you'd attract talent willing to do it faster. That's a big If depending on whether users are  profiting from using your project and are willing to support it? What's your take on this? I have no idea. My feeling was it's always better for companies to do it instead. Is that because it'll simply be the same as companies hiring people that Github has decided to go this way instead?

**Comment by jsotuyod** [comment]
> Maybe the rule wanted to prevent using "style"? Unless it wanted to prevent that html is directly used and every rendering should be done through jsf components (or similar alternatives), but than the rule should have been named "NoHtml"...

I think this is actually trying to avoid abuse… When an ad-hoc attribute is set on a tag, if it's not interpreted by the tag implementation it's usually passed through as-is to the rendered HTML (but the tag can change this behavior). The `styleClass` attribute on the other hand is supported by the renderer itself, and is therefore more reliable.

Moreover, tags, such as the one cited, may give a different meaning to the `class` attribute, raising the importance of using `styleClass` instead.

In hindsight, this was probably a very poor choice when designing the JSF tags… but that ship has already sailed.

We should whitelist tags that define a `class` attribute (and assume the user knows it's intent). Namespaces are maybe too broad + can be set arbitrarily… The definitive solution would be using the taglib uri to infer the used prefix, and then checking for that prefix + tag, but that's going to be hard on XPath… In the meantime, simply whitelisting tag names or "standard prefix" + tagname should suffice.

**Comment by Fernal73** [comment]
http://ibmwcsworld.blogspot.com/2012/09/wcbaseusebean-and-jsp-best-practises.html?m=1

Classname would have been a better choice, but, as you rightly pointed out, that ship has sailed.

**Comment by Fernal73** [comment]
This rule has been around for quite a while, yet there does not appear to be any issue raised about this rule.

Possible reasons could be:

https://stackoverflow.com/questions/18620840/what-should-i-use-instead-of-jspusebean

https://www.icesoft.org/JForum/posts/list/7370.page

How about a rule or message that discourages use of jsp:usebean in the view, as it ought to be part of the business logic instead?
